service/storage_service.cc

```diff
@@ -773,20 +797,52 @@ void storage_service::join_token_ring(int delay) {
         MigrationManager.announceNewKeyspace(TraceKeyspace.definition(), 0, false);
 #endif
 
-    // At this point our local tokens are chosen (_bootstrap_tokens) and will not be changed unless we bootstrap again.
-    db::system_keyspace::set_bootstrap_state(db::system_keyspace::bootstrap_state::COMPLETED).get();
-
     slogger.debug("Setting tokens to {}", _bootstrap_tokens);
     // This node must know about its chosen tokens before other nodes do
     // since they may start sending writes to this node after it gossips status = NORMAL.
     // Therefore, in case we haven't updated _token_metadata with our tokens yet, do it now.
     _token_metadata.update_normal_tokens(_bootstrap_tokens, get_broadcast_address());
 
+    if (!_streams_ts && db().local().get_config().check_experimental(db::experimental_features_t::CDC)) {
+        // If we didn't choose a streams timestamp at this point, then either
+        // 1. we already bootstrapped, but are upgrading from a non-CDC version (or previous version of CDC),
+        // 2. or we didn't, but we're skipping the streaming phase (seed node/auto_bootstrap=off)
+        //    and directly joining the token ring.
+
+        // In the first case (updating), we'll propose the first generation
+        // if we conclude that we're the node responsible for this.
+        // In the second case we'll always propose, just as normally bootstrapping nodes do.
+
+        // TODO: if we're replacing, we also shouldn't start a new generation.
+        // But if we're replacing, _streams_ts should be set at this point.
+        // assert(!db().local().is_replacing());
+
+        if (!db::system_keyspace::bootstrap_complete() /* second case */
+                || should_propose_first_cdc_generation()) {
+
+            _streams_ts = make_new_cdc_generation();
+        }
+    }
+
+    // Persist the streams timestamp before we persist bootstrap_state = COMPLETED.
+    if (_streams_ts) {
+        db::system_keyspace::update_streams_timestamp(*_streams_ts).get();
+    }
+    // If we crash now, we will choose a new streams timestamp anyway (because we will also choose a new set of tokens).
+    // But if we crash after setting bootstrap_state = COMPLETED, we will keep using the persisted streams timestamp after restarting.
+
```

Comment:
As @nyh mentioned, those lines are cdc specific. It is better to move to cdc.cc. This makes the changes to storage_service.cc  minimum.

Reply:
I will copy-paste my answer to nyh's comment:

If you take out the comments, there is very little code.
The "meat" of CDC code is in CDC-specific functions like make_new_cdc_generation. I will move their implementations to a different file in the cdc/ directory.

But this particular piece of code should be inlined here in join_token_ring, because to understand what it does and why it's here, you need to understand join_token_ring (and consequently, prepare_to_join and init_server... you probably need to understand 2/3 of storage_service.cc to understand how nodes start). There's a bunch of "invisible" relationships and invariants that must hold while this particular piece of code runs.

The comments are there to explain these relationships (I wish there were many more such comments in other parts of join_token_ring and prepare_to_join).

If I were to move this code to a different file, I would have to write more explanations ("this procedure must be called exactly from that place or it won't work"). And then to understand it you would have to open both storage_service.cc and the other file in a vertical split and read them consecutively.

Reply:
This is very fragile. Anyone who touches this area and isn't an expert in cdc will be afraid to touch anything.

Perhaps it's possible to name this code point? Perhaps "bootstrap_before_commit" and "bootstrap_after_commit" (no idea if those are good names, just throwing out an idea). cdc can then hook into those names. Because the names are not related to cdc, the code can be maintained without knowledge of cdc as long as the meaning is documented and understood.

Reply:
Well, I had to read storage_service.cc for a couple of days to understand the node startup code and how to modify it without breaking it (by analyzing it a bit every day, then letting my brain digest this "fantastic" piece of architecture during sleep). It's just a complicated algorithm with lots of invariants.

It is what it is, and I think anyone who'll want to modify these parts will have to go through this process.

The real fix, if possible at all, would be a complete overhaul/refactor of this set of procedures (`init_server`, `prepare_to_join`, `join_token_ring`). Maybe adding subprocedures like `bootstrap_before_commit` would help, if they had a well-specified set of clear preconditions and postconditions. But getting it right would require more work (mental work: just sitting with a piece of paper and thinking about it, not just writing code), unrelated to this patchset.

So in general I agree this needs to be changed because it's too fragile, but I don't think the change belongs to this patchset. I only scratched the tip of the mountain some time ago (https://github.com/scylladb/scylla/pull/5185). @asias what do you think?